### ISSUE TITLE
Simplify token ui and data-model

### ISSUE THREAD

**Issue by cpytel** [jdbann/table#60]
The distinction between creatures and characters isn't getting us a whole lot right now, and the data model is unnecessarily complex.

**Comment by jdbann** [comment]
What would your pitch be?

**Comment by cpytel** [comment]
Here's my pitch for how to change this, looking forward to seeing yours!

I would propose removing the `Character` and `Creature` classes, and any distinction in the UI, altogether, and just having `Token`.

I would change whether a token should be included in all maps to be a boolean on Token that you check when you create the token.

I would make it so that you can hold option/alt while dragging any token (either from the drawer or on the map) to copy it instead of move it.

Right-clicking on an individual token would show a menu that allows you to delete the token, or choose Edit, which brings up a modal to edit name, identifier, and image of that token.

The library would still be available for you to quickly bring in any token from other maps into the current map.

Here is a quick mockup of what the token creation would look like in this scenario. Additional options, like a dropdown for token size, would be added to the new token modal.

<img width="1322" alt="Screen Shot 2020-05-10 at 7 22 57 AM" src="https://user-images.githubusercontent.com/5015/81497944-e3ea8e80-928f-11ea-867c-574fd21899cb.png">

Other notes on data model:

* Token copies would all share the same image attachment (but you can edit the image on any particular token)
* `Token` would belong_to `parent` which would point to the `Token` it was copied from. This way copies can be linked together, as well as deleted all at once.

**Comment by jdbann** [comment]
My mental model of all this is very similar to the physical objects you'd have on a table:

- `Character` tokens are often the fanciest and most carefully selected.
  - Belong to the player
  - Exist on all maps
  - Player might have multiple of these (eg Ranger's companion)
- `Creature` tokens are often a pile of generic instances of a particular monster.
  - Belong to the DM
  - Exist only on maps they're needed in
  - Need to differentiate between them (Orc 1, Orc 2 etc)
- `Generic` tokens are notes of additional entities on a map.
  - Often a scrap of paper or some other makeshift token
  - Might be non-hostile creatures like Daisy the cow
  - Could belong to the players or the DM

This is where my `TokenTemplate` thought comes in. They should all have everything required to generate a token, so when we add `size` then they will all need it.

At a later date, we may have players create accounts so that they can edit their `Character` tokens, or possibly limit control of particular tokens to particular users.

At this point the token drawer would probably change. Instead of storing used tokens, it would be suggestions of tokens to use. For a new map, it would have all the `Character` templates for which a token has not been placed. After a `Creature` has been placed, it would include that creature and other creatures which have been placed on maps with this creature.

It would effectively be a dynamic view of the whole library, sorted and filtered based on what would be most useful. There would be a quick search to filter the library further (think Alfred or Basecamp's cmd+J).

There would be a clear distinction of `Token`s on the map and `Template`s in the library.

|Class|Responsibility|
|-|-|
|`Template`|Abstract class has all the validation and logic required for describing a new `Token`|
|`Character`|Inherits from `Template`, ensures only one `Token` per `Map`, possibly belongs to a player|
|`Creature`|Inherits from `Template`, adds logic for generating identifiers when multiple `Token`s are added to a `Map`|
|`Token`|A join between a descendant of `Template` and a `Map` with the token's identifier and location|

I'm not sure about `Generic` templates yet. The name is awful. They could be `Token`s which don't have an associated `Template` but I'd rather keep a consistent approach.

The type of each token's template can then be used to differentiate which behaviour applies. So rather than a DM needing to assign the particular behaviour for each token, they are decided by what was created.

`Character` templates are probably created in a `Campaign` edit flow. `Creature` templates might be created on-the-fly or in a separate campaign library. `Generic` templates will be created in the map view.

**Comment by cpytel** [comment]
I think what you're pitching is fine, and I'd be happy moving forward with it.

I agree the name Generic is awful. Maybe we just don't do that now. Let's do it as a separate feature later once we need it?

> Character templates are probably created in a Campaign edit flow. Creature templates might be created on-the-fly or in a separate campaign library. Generic templates will be created in the map view.

When I was putting what we had together, I felt like it wasn't going to be intuitive to need to go to edit a campaign to add characters, since I figured what would be motivating people is "I want to add a character to this map". That being said, its so rare that you should actually need to add a character, if it simplifies the map/token UI to push characters out, I think its worthwhile to try.

**Comment by cpytel** [comment]
Would you like to work on this and what your timeline? I'm thinking about trying to do #57 before our next session and I'm wondering if I should hold off.

**Comment by jdbann** [comment]
I'll probably not be doing anything substantial before our next session. Going to focus on prep instead. I'll try pull this apart into smaller steps in a new milestone though so we're not dealing with a long running feature branch.

**Comment by jdbann** [comment]
@cpytel What do you think about this as an approach over STI? https://github.com/rails/rails/pull/39341/files

**Comment by cpytel** [comment]
This is what I was going for with the approach I was using, so I'd be fine with it. That being said, I also think STI would be OK in this case.

**Comment by cpytel** [comment]
Naming thing:

Instead of calling the base class `Template` and the join between a descendant of `Template` and a `Map` with the token's identifier and location, `Token`, what if the base Class were Token, and the join between a descendant of `Token` and a `Map` is something like `Position` or `Placement`?

**Comment by jdbann** [comment]
The distinction I have in my head is that the `Template` is the source file for printing a `Token`. So an individual `Token` has a physical presence on the map or in the drawer, an identifier if it's one of several creatures, and potentially a detail of who's currently interacting with it.

**Comment by cpytel** [comment]
Yes, while I think `Token` and `Placement` would also work well for that, I’d prefer to just go with what is comfortable to you at this point.
